### PR TITLE
Add new models (Neural Networks, Ridge Regression, Gradient Boosting) + Ensemble method

### LINKED ISSUE #53: Replace the bagging trees with a single NN model for MLacc
To improve the MLacc performance, a neural network could be implemented to accept all input variables and predict multiple response variables at the same time. This requires modifying `ML.py` and `train.py`, combining all the training loops into a single NN training process.

### PR DESCRIPTION
**Pull request by tztsai**
In order to implement a neural network to replace the bagging trees for MLacc, we need to
- [x] For each, `ipool`, collect all input datasets into a single dataset for NN training
- [x] Restore the time dimension of the dataset with a monthly frequency to increase the data size
- [x] Implement and train an NN for each `ipool`; tune the hyperparameters using a validation dataset
- [x] Test the evaluation performance and compare it with the previous ML model

Fixes #53 #54 #57

### CODE REVIEW

**Comment by ma595** [comment]
This needs a bit of a clean up / rebase.

File: Tools/mapGlobe.py [hunk 1]
```diff
@@ -27,21 +27,20 @@
...
 ##@retval      Pred_Y                 predicted map of target variables, without masking
-def extrp_global(packdata, ipft, PFTmask, XVarName, Tree_Ens, colum, Nm, labx):
+def extrp_global(packdata, PFTmask, XVarName, Tree_Ens, colum, Nm, labx):
     global_X_map = np.full((len(XVarName), packdata.nlat, packdata.nlon), np.nan)
     # PFTmask[np.isnan(PFTmask)]=0
-    pmask = np.squeeze(PFTmask[ipft - 1][:])
     Pred_Y = np.full(PFTmask[0].shape, np.nan)
     # global metrics -> dataframe
     for ii in range(len(XVarName)):
         if ii < packdata.Nv_nopft:
```
Comment (ma595) [inline Tools/mapGlobe.py:36]: The code currently fails on this line as `packdata.Nv_nopft` has been removed from `main.py`. This was originally defined to be the `len(var_pred_name1` which corresponds to `varlist["pred"]["allname"]` which itself is ` "allname" : ["Tair", "Rainf", "Snowf", "Qair", "PSurf", "SWdown", "LWdown", "clay_frac",
             "interx"],`

Comment (ma595) [reply Tools/mapGlobe.py:36]: At this point I am quite confused. 

We encounter variables `NPP0` and `LAI0`, which are Net Primary Productivity and Leaf Area Index. They both have dimensions: `veget, lat, lon`, according to packdata. 


`packdata[XVarName[ii]` has dimensio

File: Tools/ML.py [hunk 2]
```diff
@@ -383,83 +243,80 @@ def MLloop(
...
-                            res["var"] = varname
-                            for i, (k, v) in enumerate(dim_ind):
-                                res[f"dim_{i+1}"] = k
-                                res[f"ind_{i+1}"] = v
-                            result.append(res)
-                # close&save netCDF file
-                restnc.close()
-
-    return pd.DataFrame(result).set_index("var")
+                        break
```
Comment (ma595) [inline Tools/ML.py:294]: Is there a reason we break here?

File: Tools/ML.py [hunk 3]
```diff
@@ -383,83 +243,80 @@ def MLloop(
...
+
+                    # close&save netCDF file
+                    restnc.close()
+
+                    if len(comb_ds[ipool]) > 3:
+                        break
```
Comment (ma595) [inline Tools/ML.py:300]: Is there a reason we break here also?

File: Tools/ML.py (hunk 2 above)
Comment (ma595) [reply Tools/ML.py:294]: I assume this was for debugging purposes?

File: Tools/mapGlobe.py
Comment (ma595) [inline Tools/mapGlobe.py:1]: I understand the shape of `Ym` and `Ymm`. However, the rows are all identical (easier to see in `Ym`. Does this make sense? For biomass specifically (I think):

```python
duplicates = Ym.duplicated(keep=False)
duplicate_counts = Ym[duplicates].value_counts()
print(duplicate_counts)

(4194)
```

**Review by ma595 — COMMENTED** [review]
@tztsai Checking your changes from last week. Can you please clarify for my benefit. Thanks.

File: Tools/ML.py [hunk 4]
```diff
@@ -13,223 +13,108 @@
...
+    Combine data from multiple dictionaries into a single DataFrame.
+
+    Args:
+        dics (dict): Dictionary of data to combine.
+
+    Returns:
+        pandas.DataFrame: Combined data with NaN values and 'pft' column dropped.
+    """
+    df = pd.DataFrame(dics)
+    df = df.dropna().drop(columns=['pft'])  # drop data according to the PFT mask
```
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/ML.py:86]: This is equivalent to the original code (dropping the rows with pft=NaN), but we can try reversing the order.

[the remark this answers is not in the document]
Comment (ma595) [reply Tools/ML.py:86]: Am I missing something here? If you print the pft map (non-nan values) for each pft (biomass):

| Column 1 | Column 2 |
|----------|----------|
| 2        | 102      |
| 3        | 124      |
| 4        | 84       |
| 5        | 59       |
| 6        | 111      |
| ...      | ...      |
| 14       | 65       |
| 15       | 268      |

[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/ML.py:86]: I see. That's a good point. Thanks for pointing it out!

[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/ML.py:86]: I've updated the code using your peice of code now.

File: Tools/mapGlobe.py
Comment (tztsai) [reply Tools/mapGlobe.py:1]: Can you indicate which `ipool` and `varname` it belongs to?

Comment (ma595) [reply Tools/mapGlobe.py:1]: Looks like they are all different now. Maybe it was fixed in the last commit?

Comment (tztsai) [reply Tools/mapGlobe.py:1]: OK, thanks for checking

**Comment by ma595** [comment]
@tztsai This is really good work. I will run the code and review over the next few days. 

I would suggest adding some comments in the code to provide a high level summary of what some of the routines are doing. Some more comments in the code itself would be helpful too. I am happy to take some of this on as I review. 

It would also be useful to have an automated test to ensure that regressions to the data pipeline have not been introduced: 
- Show that the same dataset is being generated at the start of training as is generated in the old code (main).
- Do tell me if this doesn't make sense, for instance due to any stochasticity in the pipeline. 

Because of the size of this PR and the fact it touches so many files, it might be good to split up into two:
- Modified data pipeline and implement test to check results are same as before (if relevant)
- Additional ML models, tuning and parallelisation. 

On the ML side, could you provide a summary of how the performance changes? I noticed you made a CHANGES.txt file. Possibly summarise the improvements to ML performance there for now. 

I looked at increasing the size of Ncc as well and it seems that the increase you first tried was beyond the size of the available data. In the CHANGES.txt it still suggests that the Nc size was increase. Could this please be corrected.

**Comment by ma595** [comment]
https://github.com/CALIPSO-project/SPINacc/pull/54/commits/d6ea9101c4083a30e89147e76f5fe779c68b1a7e 
The message associated with this commit does not fully reflect what has been changed.

**Comment by tztsai** [comment]
@ma595 Thanks for the review! Would you be able to help splitting the PR, in the way you suggested?

**Comment by ma595** [comment]
> @ma595 Thanks for the review! Would you be able to help splitting the PR, in the way you suggested? I have added an automated test to make sure that the ML performance is better than the previous run, but I will add another test to check that the dataset does not change as well.

Hi @tztsai, happy to help splitting the PR.

File: Tools/readvar.py [hunk 5]
```diff
@@ -59,7 +66,7 @@ def readvar(varlist, config, logfile):
             if "land" in f[varname_clim[index]].dimensions:
                 land = f["land"][:] - 1
                 ntime = len(da)
-                uncomp = np.ma.masked_all((ntime, nlat * nlon))
+                uncomp = np.arr.masked_all((ntime, nlat * nlon))
```
Comment (ma595) [inline Tools/readvar.py:69]: Does this line make sense? I haven't heard of `np.arr` before.

**Comment by ma595** [comment]
The performance is starting to look good, well done @tztsai. Just a comment on splitting up:

I plan to split it into 2 or 3 PRs:
- 1. Testing #45 
- 2. Pipeline refactoring (currently not opened). 
- 3. New ML implementations and parameter tuning. 

I am unsure whether to split into 2 and 3 as the refactoring and new ML implementation are quite tightly coupled.

File: Tools/readvar.py (hunk 5 above)
Comment (ma595) [reply Tools/readvar.py:69]: @tztsai bump

Comment (tztsai) [reply Tools/readvar.py:69]: I think it does not make sense

**Comment by ma595** [comment]
OK, as this PR has been sitting here for a few months, I have decided to squash the changes instead of rebasing and cleaning up in the interest of time. I'll use the CHANGES.md as the basis of the squash.

File: CHANGES.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# Summary of Changes
+
+- Converted dataset format to xarray.Dataset - merged auxil into packdata, etc.
+- Reformat documentation and added more comments
+- Rewrote tests as pytest files
+- Reformated the config file as a python file
+- Used ruff for reformatting and linting
+- Added pre-commit checks and a Github workflow for CI checks
+- Added input features (std of some variables like Qair, Psurf, etc.)
```
[the remark this answers is not in the document]
Comment (tztsai) [reply CHANGES.md:9]: The original code did not take std of these variables, and I added them

File: Tools/ML.py
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/ML.py:1]: Yes this function was never used

File: Tools/extract_X.py
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/extract_X.py:1]: It could be written in this way even in the original format (numpy array)

File: main.py [hunk 7]
```diff
@@ -207,22 +210,39 @@
...
+                labx,
+                config,
+                restfile,
+                alg,
+            )
+            result.append(res_df)
+        res_df = pd.concat(result, keys=Yvar.keys(), names=["comp"])
+        print(res_df)
+
+        scores = res_df.mean()[["R2", "slope"]].to_frame().T
```
[the remark this answers is not in the document]
Comment (tztsai) [reply main.py:231]: Yes, but the MLacc_results.csv format was also changed (including a new column indicating the `alg`, aggregating scores, etc.)

File: Tools/train.py
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/train.py:1]: It seems xgboost is always better than random forest. Since they are in the same category, it seems that RF is unnecessary.

File: Tools/ML.py [hunk 8]
```diff
@@ -252,22 +121,21 @@ def MLmap_multidim(
...
+        predY = np.where(pool_map == pool_map, predY_train.iloc[0], np.nan)
         Global_Predicted_Y_map = predY
     else:
         Global_Predicted_Y_map, predY = mapGlobe.extrp_global(
             packdata,
             ipft,
             PFT_mask,
-            var_pred_name,
-            Tree_Ens,
+            combine_XY.columns.drop("Y"),
```
[the remark this answers is not in the document]
Comment (tztsai) [reply Tools/ML.py:135]: Previously, `var_pred_name` was hand-coded as a list of column names. Now that we have changed the input features in `readvar.py` (added a few new features), it needs to be updated as `combine_XY.columns.drop("Y")`

**Comment by ma595** [comment]
A few updates:

- I've rebased changes from main and am working through the changes
- First observation after rebasing is that none of the tests currently pass:
  - `Tamp` was excluded from packdata (causing `test_task1.py` to fail.)
  - Was there any justification to do this @tztsai? 
  - All temperature statistics were incorrect because `Tair` was being overwritten after converting to celsius
  - After these changes `test_init.py`, `test_task1.py` and `test_task2.py` now pass. 
- `test_task4.py` still fails - which isn't a big surprise
  - Debugging strategy was to output the combine_XY between main and nn branches. Obviously differences emerge:
  - We conclude that this is because there are some additional variables:
  - `[LWdown_std', 'PSurf_std', 'Qair_std', 'SWdown_std', 'Snowf_mean','Snowf_std']`
  - All other columns match between old and new.
